From 3633aa500fafdf5cdc221a82128889e28d436e26 Mon Sep 17 00:00:00 2001
From: Anson Huang <Anson.Huang@nxp.com>
Date: Mon, 25 Jan 2016 22:16:48 +0800
Subject: [PATCH 3638/5242] MLK-12262-1 ARM: imx: enable ddr auto self-refresh
 for i.MX7D

commit  6a2fbff6f8ac9b880a8a05c7e71871f89c0609c0 from
https://source.codeaurora.org/external/imx/linux-imx.git

Enable DDR auto self-refresh for i.MX7D, when doing DDR
frequency scale or suspend/resume, DDR self-refresh will
be disabled, this is incorrect for saving power, enable it
for all these scenarios.

Signed-off-by: Anson Huang <Anson.Huang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/mach-imx/ddr3_freq_imx7d.S |   10 ++++++++++
 arch/arm/mach-imx/lpddr3_freq_imx.S |    5 +++++
 arch/arm/mach-imx/suspend-imx7.S    |    5 +++++
 3 files changed, 20 insertions(+)

diff --git a/arch/arm/mach-imx/ddr3_freq_imx7d.S b/arch/arm/mach-imx/ddr3_freq_imx7d.S
index 7e4b49d..8aa219f 100644
--- a/arch/arm/mach-imx/ddr3_freq_imx7d.S
+++ b/arch/arm/mach-imx/ddr3_freq_imx7d.S
@@ -237,6 +237,11 @@
 	ldr	r7, =0x0
 	str	r7, [r4, #DDRC_DBG1]
 
+	/* enable auto self-refresh */
+	ldr	r7, [r4, #DDRC_PWRCTL]
+	orr	r7, r7, #(1 << 0)
+	str	r7, [r4, #DDRC_PWRCTL]
+
 	.endm
 
 	.macro	switch_to_533m
@@ -433,6 +438,11 @@
 	ldr	r7, =0x0
 	str	r7, [r4, #DDRC_DBG1]
 
+	/* enable auto self-refresh */
+	ldr	r7, [r4, #DDRC_PWRCTL]
+	orr	r7, r7, #(1 << 0)
+	str	r7, [r4, #DDRC_PWRCTL]
+
 	.endm
 
 ENTRY(imx7d_ddr3_freq_change)
diff --git a/arch/arm/mach-imx/lpddr3_freq_imx.S b/arch/arm/mach-imx/lpddr3_freq_imx.S
index a33b893..80fb118 100644
--- a/arch/arm/mach-imx/lpddr3_freq_imx.S
+++ b/arch/arm/mach-imx/lpddr3_freq_imx.S
@@ -118,6 +118,11 @@
 	ldr	r7, =0x1
 	str	r7, [r4, #DDRC_PCTRL_0]
 
+	/* enable auto self-refresh */
+	ldr	r7, [r4, #DDRC_PWRCTL]
+	orr	r7, r7, #(1 << 0)
+	str	r7, [r4, #DDRC_PWRCTL]
+
 	.endm
 
 	.macro	switch_to_below_100m
diff --git a/arch/arm/mach-imx/suspend-imx7.S b/arch/arm/mach-imx/suspend-imx7.S
index 744621e..91019802 100644
--- a/arch/arm/mach-imx/suspend-imx7.S
+++ b/arch/arm/mach-imx/suspend-imx7.S
@@ -490,6 +490,11 @@
 	ldr	r7, =0x1
 	str	r7, [r3, #DDRC_PCTRL_0]
 
+	/* enable auto self-refresh */
+	ldr	r7, [r3, #DDRC_PWRCTL]
+	orr	r7, r7, #(1 << 0)
+	str	r7, [r3, #DDRC_PWRCTL]
+
 	.endm
 
 ENTRY(imx7_suspend)
-- 
1.7.9.5

